protobuf_serialization/writer.nim

```diff
@@ -0,0 +1,251 @@
+
+#Create a field key.
+template key(fieldNum: uint, wire: ProtobufWireType): byte =
+  ((byte(fieldNum shl 3)) or wire.byte).byte
+
+#Created in response to https://github.com/kayabaNerve/nim-protobuf-serialization/issues/5.
+proc verifyWritable[T](ty: typedesc[T]) {.compileTime.} =
+  when T is PlatformDependentTypes:
+    {.fatal: "Writing a number requires specifying the amount of bits via the type.".}
+  elif T is ((PureSIntegerTypes or PureUIntegerTypes) and (not bool)):
+    {.fatal: "Writing a number requires specifying the encoding to use.".}
+  elif T.isStdlib():
+    discard
+  elif T is (object or ref):
+    enumInstanceSerializedFields(T(), fieldName, fieldVar):
+      discard fieldName
+      when fieldVar is PlatformDependentTypes:
+        {.fatal: "Writing a number requires specifying the amount of bits via the type.".}
+      elif fieldVar is (VarIntTypes or FixedTypes):
+        const
+          hasPInt = ty.hasCustomPragmaFixed(fieldName, pint)
+          hasSInt = ty.hasCustomPragmaFixed(fieldName, sint)
+          hasUInt = (ty.hasCustomPragmaFixed(fieldName, puint) or (flatType(fieldVar) is bool))
+          hasFixed = ty.hasCustomPragmaFixed(fieldName, fixed)
+        when uint(hasPInt) + uint(hasSInt) + uint(hasUInt) + uint(hasFixed) != 1:
+          {.fatal: "Couldn't write " & fieldName & "; either none or multiple encodings were specified.".}
+
+    if totalSerializedFields(T) > 32:
+      raise newException(Defect, "Object has too many fields; Protobuf has a maximum of 32.")
+
+proc writeVarInt(
+  stream: OutputStream,
+  fieldNum: uint,
+  value: VarIntWrapped
+) {.raises: [Defect, IOError].} =
+  let bytes = encodeVarInt(value)
+  if (bytes.len == 1) and (bytes[0] == 0):
+    return
+  stream.write(key(fieldNum, VarInt))
+  stream.write(bytes)
+
+proc writeFixed(
+  stream: OutputStream,
+  fieldNum: uint,
+  value: FixedWrapped
+) {.raises: [Defect, IOError].} =
+  when sizeof(value) == 8:
+    var raw = cast[uint64](value)
+  else:
+    var raw = cast[uint32](value)
+  if raw == 0:
+    return
+  stream.write(key(
+    fieldNum,
+    when sizeof(value) == 8:
+      Fixed64
+    else:
+      Fixed32
+  ))
+  for _ in 0 ..< sizeof(value):
+    stream.write(byte(raw and LAST_BYTE))
+    raw = raw shr 8
+
+#stdlib types toProtobuf's. inlined as it needs access to the writeValue function.
+include stdlib_writers
+
+proc writeValueInternal[T](
+  stream: OutputStream,
+  value: T,
+  sub: bool,
+  existingLength: var int
+) {.raises: [Defect, IOError, ProtobufWriteError].}
+
+proc writeLengthDelimited[T](
+  stream: OutputStream,
+  fieldNum: uint,
+  rootType: typedesc[T],
+  flatValue: LengthDelimitedTypes,
+  sub: bool,
+  existingLength: var int
+) {.raises: [Defect, IOError, ProtobufWriteError].} =
+  if existingLength > 255:
+    raise newException(ProtobufWriteError, "Buffer length exceeded 255 when writing a new nested object.")
+  if sub:
+    existingLength += 2
+
+  var bytes: seq[byte]
```

Comment:
No need for buffering here. You can use the ["delayed writes"](https://github.com/status-im/nim-faststreams/#delayed-writes) feature of the output stream to write the varint value after all the other bytes are written.

There will be no temporary `memoryOutputs` here and all output functions will directly work with an `OutputStream` instead of populating a sequence first.